src/update_halo.jl

```diff
@@ -477,100 +471,67 @@ end
+            # end
         end
     end
 end
 
 let
-    global iread_recvbufs!, allocate_rocqueues_iread, wait_iread
+    global iread_recvbufs!, allocate_rocstreams_iread, wait_iread
 
-    rocqueues  = Array{AMDGPU.HSAQueue}(undef, NNEIGHBORS_PER_DIM, 0)
-    rocsignals = Array{Union{AMDGPU.HSASignal,AMDGPU.RuntimeEvent{AMDGPU.HSAStatusSignal},Missing}}(undef, NNEIGHBORS_PER_DIM, 0)
+    rocstreams = Array{AMDGPU.HIPStream}(undef, NNEIGHBORS_PER_DIM, 0)
 
-    function wait_iread(n::Integer, A::ROCArray{T}, i::Integer) where T <: GGNumber
-        if !ismissing(rocsignals[n,i]) # DEBUG: tmp solution to avoid rocsignals array access filing when accessing an unset signal
-            wait(rocsignals[n,i]);
-            rocsignals[n,i] = missing;
-        end
-        return
-    end
-
-    function allocate_rocqueues_iread(fields::GGArray...)
-        if length(fields) > size(rocqueues,2)  # Note: for simplicity, we create a stream for every field even if it is not a CuArray
-            nqueues = length(fields)-size(rocqueues,2);
-            new_rocqueues  = Array{AMDGPU.HSAQueue}(undef, NNEIGHBORS_PER_DIM, nqueues);
-            new_rocsignals = Array{Union{AMDGPU.HSASignal,AMDGPU.RuntimeEvent{AMDGPU.HSAStatusSignal},Missing}}(missing, NNEIGHBORS_PER_DIM, nqueues); # DEBUG: tmp solution to avoid rocsignals array access filing when accessing an unset signal
-            for i = 1:nqueues
-                for n=1:NNEIGHBORS_PER_DIM
-                    q = AMDGPU.HSAQueue(get_default_agent())
-                    AMDGPU.HSA.amd_queue_set_priority(q.queue, AMDGPU.HSA.AMD_QUEUE_PRIORITY_HIGH)
-                    new_rocqueues[n,i] = q
-                end
-            end
-            rocqueues  = [rocqueues  new_rocqueues]
-            rocsignals = [rocsignals new_rocsignals]
+    wait_iread(n::Integer, A::ROCArray{T}, i::Integer) where T <: GGNumber = AMDGPU.synchronize(rocstreams[n,i]);
+
+    function allocate_rocstreams_iread(fields::GGArray...)
+        if length(fields) > size(rocstreams,2)  # Note: for simplicity, we create a stream for every field even if it is not a ROCArray
+            rocstreams = [rocstreams [AMDGPU.HIPStream(:high) for n=1:NNEIGHBORS_PER_DIM, i=1:(length(fields)-size(rocstreams,2))]];  # Create (additional) maximum priority nonblocking streams to enable overlap with computation kernels.
         end
     end
 
     function iread_recvbufs!(n::Integer, dim::Integer, A::ROCArray{T}, i::Integer) where T <: GGNumber
         if ol(dim,A) >= 2  # There is only a halo and thus a halo update if the overlap is at least 2...
             # DEBUG: the follow section needs perf testing
-            if dim == 1 || amdgpuaware_MPI(dim)  # Use a custom copy kernel for the first dimension to obtain a good copy performance (the CUDA 3-D memcopy does not perform well for this extremely strided case).
-                ranges   = recvranges(n, dim, A);
+            # DEBUG 2: commenting read_h2d_async! for now
```

Comment:
For now I am commenting the async memcpy option and only rely on kernel since there is an issue most probably in the underlying HIP function.